OpenRA.Mods.Common/Traits/Player/ClassicParallelProductionQueue.cs

```diff
@@ -154,9 +164,11 @@ protected override bool BuildUnit(ActorInfo unit)
 
 			var producers = self.World.ActorsWithTrait<Production>()
 				.Where(x => x.Actor.Owner == self.Owner
-					&& !x.Trait.IsTraitDisabled
-					&& x.Trait.Info.Produces.Contains(type))
-				.OrderByDescending(x => x.Actor.IsPrimaryBuilding())
+		            && !x.Trait.IsTraitDisabled
+		            && x.Trait.Info.Produces.Contains(type))
+				.Select<TraitPair<Production>, (Actor Actor, Production Production, PrimaryBuilding PrimaryBuilding)>(
+					x => (x.Actor, x.Trait, x.Actor.TraitOrDefault<PrimaryBuilding>()))
+				.OrderByDescending(x => x.PrimaryBuilding != null && x.PrimaryBuilding.IsPrimary)
```

Comment:
Worth to extract these lines and reuse the same method in the other places?

Reply:
Yeah, this could go as a static helper on PrimaryBuilding maybe.

But all of this needs to be reevaluated anyway to change it to prioritise based on distance from primary instead of outright blocking production.